pkg/manager/config/manager.go

```diff
@@ -42,108 +43,89 @@ type KVValue struct {
 	Value []byte
 }
 
-type KVEventType byte
-
-const (
-	KVEventPut KVEventType = iota
-	KVEventDel
-)
-
-type KVEvent struct {
-	KVValue
-	Type KVEventType
-}
-
-type kvListener struct {
-	key string
-	wfn func(*zap.Logger, KVEvent)
-}
-
 type ConfigManager struct {
-	wg        waitgroup.WaitGroup
-	cancel    context.CancelFunc
-	kv        *btree.BTreeG[KVValue]
-	events    chan KVEvent
-	listeners []kvListener
-	logger    *zap.Logger
+	wg     waitgroup.WaitGroup
+	cancel context.CancelFunc
+	logger *zap.Logger
+
+	kv *btree.BTreeG[KVValue]
+
+	wch     *fsnotify.Watcher
+	overlay string
+	sts     struct {
+		sync.Mutex
+		listeners []chan<- *config.Config
+		current   *config.Config
+	}
 }
 
 func NewConfigManager() *ConfigManager {
 	return &ConfigManager{}
 }
 
-func (srv *ConfigManager) Init(ctx context.Context, cfg *config.Config, logger *zap.Logger) error {
-	srv.logger = logger
-	srv.events = make(chan KVEvent, 256)
-	srv.kv = btree.NewBTreeG(func(a, b KVValue) bool {
+func (e *ConfigManager) Init(ctx context.Context, logger *zap.Logger, configFile string, overlay *config.Config) error {
+	e.logger = logger
+
+	// for namespace persistence
+	e.kv = btree.NewBTreeG(func(a, b KVValue) bool {
 		return a.Key < b.Key
 	})
 
-	// init config for other components
-	if err := srv.SetConfig(ctx, cfg); err != nil {
-		return err
+	// for config watch
+	wch, err := fsnotify.NewWatcher()
+	if err != nil {
+		return errors.WithStack(err)
 	}
+	e.wch = wch
+	parentDir := filepath.Dir(configFile)
 
 	var nctx context.Context
-	nctx, srv.cancel = context.WithCancel(ctx)
-	srv.wg.Run(func() {
+	nctx, e.cancel = context.WithCancel(ctx)
+	e.wg.Run(func() {
+		ticker := time.NewTicker(200 * time.Millisecond)
 		for {
 			select {
 			case <-nctx.Done():
 				return
-			case ev := <-srv.events:
-				for _, lsn := range srv.listeners {
-					lsn.wfn(srv.logger, ev)
+			case err := <-e.wch.Errors:
+				e.logger.Info("failed to watch config file", zap.Error(err))
+			case ev := <-e.wch.Events:
+				e.handleFSEvent(ev, configFile)
+			case <-ticker.C:
+				if err := e.wch.Add(parentDir); err != nil {
```

Comment:
Yes, but eventually, you need to re-watch the file somehow. If I only re-watch the file on `remove/renaming`, it will sometime miss the event, and eventually stop watching.

That caused some unstable failures in tests. So I switched to dir watching for more robust events handling.

Reply:
And why repeatedly watch?

Reply:
Because directories/files may be removed completely in some cases. And before they are created again, you can't re-watch files/directories.

So I used a ticker to retry infinitely.

Reply:
Please add all the explanations to the comment, including watching directory, repeatedly watching.

Reply:
Added